File: water-system/src/main/java/com/water/water_system/controller/WaterController.java
package com.water.water_system.controller;

import com.water.water_system.entity.Tenant;
import com.water.water_system.entity.Unit;
import com.water.water_system.entity.Payment;
import com.water.water_system.service.WaterService;
import com.water.water_system.repository.UnitRepository;
import com.water.water_system.dto.BillRequestDTO;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api")
@CrossOrigin(origins = "*")
public class WaterController {

    @Autowired
    private WaterService waterService;

    @Autowired
    private UnitRepository unitRepository;

    @GetMapping("/tenants")
    public List<Tenant> getTenants() {
        return waterService.getAllTenants();
    }

    @PostMapping("/units")
    public Unit addUnits(@RequestBody Map<String, Object> payload) {
        Long tenantId = Long.valueOf(payload.get("tenantId").toString());
        String month = payload.get("month").toString();
        Integer units = Integer.valueOf(payload.get("units").toString());
        return waterService.addUnits(tenantId, month, units);
    }

    @GetMapping("/dashboard")
    public Map<String, Object> getDashboard(@RequestParam String month) {
        return waterService.getDashboardData(month);
    }
    
    @GetMapping("/usage")
    public List<Map<String, Object>> getUsage(@RequestParam String month) {
        return waterService.getTenantUsage(month);
    }

    @PostMapping("/bill")
    public void setBill(@RequestBody BillRequestDTO billRequest) {
        waterService.setBill(billRequest.getMonth(), billRequest.getTotalBill(), billRequest.getTotalUnits());
    }

    @GetMapping("/payments")
    public List<Payment> getPayments(@RequestParam String month) {
        return waterService.getPaymentsByMonth(month);
    }

    @PostMapping("/payments")
    public Payment addPayment(@RequestBody Map<String, Object> payload) {
        Long tenantId = Long.valueOf(payload.get("tenantId").toString());
        String month = payload.get("month") != null ? payload.get("month").toString() : "Current";
        BigDecimal amount = new BigDecimal(payload.get("amount").toString());
        return waterService.addPayment(tenantId, month, amount);
    }

    @GetMapping("/tenant-summary")
    public List<Map<String, Object>> getTenantSummary(@RequestParam String month) {
        return waterService.getTenantSummary(month);
    }

    @PostMapping("/register")
    public Tenant register(@RequestBody Map<String, String> payload) {
        String name = payload.get("name");
        String phone = payload.get("phone");
        return waterService.registerTenant(name, phone);
    }

    @PostMapping("/login")
    public Tenant login(@RequestBody Map<String, String> payload) {
        String phone = payload.get("phone");
        return waterService.loginByPhone(phone);
    }

    @GetMapping("/current-user")
    public Tenant getCurrentUser(@RequestParam Long tenantId) {
        return waterService.getTenantById(tenantId);
    }

    @GetMapping("/tenant-units")
    public Unit getTenantUnits(@RequestParam Long tenantId, @RequestParam String month) {
        return unitRepository.findByTenantIdAndMonth(tenantId, month).orElse(new Unit());
    }

    @PostMapping("/admin-check")
    public Map<String, Object> checkAdmin(@RequestBody Map<String, String> payload) {
        String phone = payload.get("phone");
        boolean isAdmin = waterService.isAdmin(phone);
        Map<String, Object> result = new HashMap<>();
        result.put("isAdmin", isAdmin);
        return result;
    }

    @GetMapping("/tenants/{id}")
    public Tenant getTenant(@PathVariable Long id) {
        return waterService.getTenantById(id);
    }

    @PostMapping("/tenants/create")
    public Tenant createTenant(@RequestBody Map<String, Object> payload) {
        String name = payload.get("name").toString();
        String phone = payload.get("phone").toString();
        String roleStr = payload.get("role") != null ? payload.get("role").toString() : "TENANT";
        Integer prevUnits = payload.get("previousUnits") != null ? Integer.valueOf(payload.get("previousUnits").toString()) : 0;
        String doorNumber = payload.get("doorNumber") != null ? payload.get("doorNumber").toString() : null;
        
        Tenant.Role role = "ADMIN".equalsIgnoreCase(roleStr) ? Tenant.Role.ADMIN : Tenant.Role.TENANT;
        return waterService.createTenant(name, phone, role, prevUnits, doorNumber);
    }

    @PutMapping("/tenants/{id}")
    public Tenant updateTenant(@PathVariable Long id, @RequestBody Map<String, Object> payload) {
        String name = payload.get("name").toString();
        String phone = payload.get("phone").toString();
        String roleStr = payload.get("role") != null ? payload.get("role").toString() : "TENANT";
        Integer prevUnits = payload.get("previousUnits") != null ? Integer.valueOf(payload.get("previousUnits").toString()) : 0;
        String doorNumber = payload.get("doorNumber") != null ? payload.get("doorNumber").toString() : null;
        
        Tenant.Role role = "ADMIN".equalsIgnoreCase(roleStr) ? Tenant.Role.ADMIN : Tenant.Role.TENANT;
        return waterService.updateTenant(id, name, phone, role, prevUnits, doorNumber);
    }

    @DeleteMapping("/tenants/{id}")
    public Map<String, String> deleteTenant(@PathVariable Long id) {
        try {
            waterService.deleteTenant(id);
            Map<String, String> result = new HashMap<>();
            result.put("message", "Tenant deleted successfully");
            return result;
        } catch (Exception e) {
            System.err.println("Delete Tenant Error: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Could not delete tenant: " + e.getMessage());
        }
    }

    @PostMapping("/tenants/{id}/baseline")
    public Tenant updateBaseline(@PathVariable Long id, @RequestBody Map<String, Object> payload) {
        try {
            Object unitsObj = payload.get("previousUnits");
            Integer units = unitsObj != null ? Integer.valueOf(unitsObj.toString()) : 0;
            return waterService.updateTenantBaseline(id, units);
        } catch (Exception e) {
            System.err.println("Baseline Update Error: " + e.getMessage());
            e.printStackTrace();
            throw new RuntimeException("Failed to update baseline: " + e.getMessage());
        }
    }
}

File: water-backend/app/Http/Controllers/Api/WaterController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Services\WaterService;
use App\Models\Unit;
use Illuminate\Http\Request;

class WaterController extends Controller
{
    protected $waterService;

    public function __construct(WaterService $waterService)
    {
        $this->waterService = $waterService;
    }

    public function getTenants()
    {
        return response()->json($this->waterService->getAllTenants());
    }

    public function addUnits(Request $request)
    {
        $tenantId = $request->input('tenantId');
        $month = $request->input('month');
        $units = $request->input('units');

        $unit = $this->waterService->addUnits($tenantId, $month, $units);
        return response()->json($unit);
    }

    public function getDashboard(Request $request)
    {
        $month = $request->query('month');
        return response()->json($this->waterService->getDashboardData($month));
    }

    public function getUsage(Request $request)
    {
        $month = $request->query('month');
        return response()->json($this->waterService->getTenantUsage($month));
    }

    public function setBill(Request $request)
    {
        $month = $request->input('month');
        $totalBill = $request->input('totalBill');
        $totalUnits = $request->input('totalUnits');
        
        $this->waterService->setBill($month, $totalBill, $totalUnits);
        
        return response()->json(['message' => 'Bill set successfully']);
    }

    public function getPayments(Request $request)
    {
        $month = $request->query('month');
        return response()->json($this->waterService->getPaymentsByMonth($month));
    }

    public function addPayment(Request $request)
    {
        $tenantId = $request->input('tenantId');
        $month = $request->input('month', 'Current');
        $amount = $request->input('amount');
        
        $payment = $this->waterService->addPayment($tenantId, $month, $amount);
        return response()->json($payment);
    }

    public function getTenantSummary(Request $request)
    {
        $month = $request->query('month');
        return response()->json($this->waterService->getTenantSummary($month));
    }

    public function register(Request $request)
    {
        $name = $request->input('name');
        $phone = $request->input('phone');
        
        $tenant = $this->waterService->registerTenant($name, $phone);
        return response()->json($tenant);
    }

    public function login(Request $request)
    {
        $phone = $request->input('phone');
        $tenant = $this->waterService->loginByPhone($phone);
        return response()->json($tenant);
    }

    public function getCurrentUser(Request $request)
    {
        $tenantId = $request->query('tenantId');
        return response()->json($this->waterService->getTenantById($tenantId));
    }

    public function getTenantUnits(Request $request)
    {
        $tenantId = $request->query('tenantId');
        $month = $request->query('month');
        
        $unit = Unit::where('tenant_id', $tenantId)->where('month', $month)->first();
        if (!$unit) {
            // Spring Boot returned empty Unit
            return response()->json(new \stdClass());
        }
        return response()->json($unit);
    }

    public function checkAdmin(Request $request)
    {
        $phone = $request->input('phone');
        $isAdmin = $this->waterService->isAdmin($phone);
        return response()->json(['isAdmin' => $isAdmin]);
    }

    public function getTenant($id)
    {
        return response()->json($this->waterService->getTenantById($id));
    }

    public function createTenant(Request $request)
    {
        $name = $request->input('name');
        $phone = $request->input('phone');
        $roleStr = $request->input('role', 'TENANT');
        $prevUnits = $request->input('previousUnits', 0);
        $doorNumber = $request->input('doorNumber');

        $role = strtoupper($roleStr) === 'ADMIN' ? 'ADMIN' : 'TENANT';
        
        $tenant = $this->waterService->createTenant($name, $phone, $role, $prevUnits, $doorNumber);
        return response()->json($tenant);
    }

    public function updateTenant(Request $request, $id)
    {
        $name = $request->input('name');
        $phone = $request->input('phone');
        $roleStr = $request->input('role', 'TENANT');
        $prevUnits = $request->input('previousUnits', 0);
        $doorNumber = $request->input('doorNumber');

        $role = strtoupper($roleStr) === 'ADMIN' ? 'ADMIN' : 'TENANT';

        $tenant = $this->waterService->updateTenant($id, $name, $phone, $role, $prevUnits, $doorNumber);
        return response()->json($tenant);
    }

    public function deleteTenant($id)
    {
        try {
            $this->waterService->deleteTenant($id);
            return response()->json(['message' => 'Tenant deleted successfully']);
        } catch (\Exception $e) {
            return response()->json(['error' => 'Could not delete tenant: ' . $e->getMessage()], 500);
        }
    }

    public function updateBaseline(Request $request, $id)
    {
        try {
            $units = $request->input('previousUnits', 0);
            $tenant = $this->waterService->updateTenantBaseline($id, $units);
            return response()->json($tenant);
        } catch (\Exception $e) {
            return response()->json(['error' => 'Failed to update baseline: ' . $e->getMessage()], 500);
        }
    }
}
